### ISSUE TITLE
[Optimisation] Transactions

### ISSUE THREAD

**Issue by CitrusWire** [2shady4u/godot-sqlite#27]
I'm not into C++, but looking at the code, it doesn't seem like you're using transactions.
Certainly I'd suggest wrapping `insert_rows`, `update_rows`, `delete_rows` inside a transaction:

https://www.sqlite.org/lang_transaction.html

Just park a BEGIN in front of it and a COMMIT after it. Use ROLLBACK for if any of them fail.
Otherwise SQLite autocommits after every statement; this speeds it up - https://www.sqlite.org/faq.html#q19
(I'm not using those methods myself, but others may benefit).

**Comment by 2shady4u** [comment]
Hello @CitrusWire 

This is a good suggestion.
At the moment you can use following work-around:

```
db.query("BEGIN TRANSACTION;")
# Call insert_rows(), update_rows() or delete_rows() here
db.insert_rows("some_table", [])
db.query("END TRANSACTION;")
```

There might be cases when you DONT want to all of your inserts to be ONE big commit... 
But I guess in that case you could then just use a for-loop in gdscript and call insert_row() one a a time?

I'll have to think about this one, since there is an easy work-around and having this behaviour by default might not be the right way to go. 🤔

**Comment by CitrusWire** [comment]
Yep, definitely easy to work around, but most people don't have the level of SQLite knowledge to know about it (especially not game devs) and it's an easy optimisation win, so personally I'd suggest this be the default.

I'm not sure I've ever come across any scenarios where I didn't want all my stuff to be ran at once. That's generally the preferred way you want databases to run and it's a core tenet of ATOMIC.
The exception I guess is how you want error handling to work. In transactions if something goes wrong then the entire thing doesn't commit; that's usually a desirable outcome (so you don't end up with a half-updated/inserted/deleted database!), but there are probably scenarios where you may not want that.

My suggestion: Enable this, document it (including a bit about how if any part of the transaction works, no part of the transaction is applied, because that will be a gotcha for non-database types), and say "if you don't want this, use singular-named functions in a `for` loop".

**Comment by 2shady4u** [comment]
Did some small tests to see the difference in performance. All with inserting 1000 rows with random integers as data.

These are the results:

**A. insert_row() WITHOUT transactions:**

```gdscript
var time := OS.get_system_time_msecs()
for i in range(0, NUMBER_OF_ROWS):
    db.insert_row(table_name, {"id": i, "number": randi()})
print(OS.get_system_time_msecs() - time)
```

elapsed times (tested 3 times):
19215 ms
23271 ms
22336 ms 

**B. insert_rows() WITHOUT transactions:**

```gdscript
var time := OS.get_system_time_msecs()
var row_array := []
for i in range(0, NUMBER_OF_ROWS):
    row_array.append({"id": i, "number": randi()})
db.insert_rows(table_name, row_array)
print(OS.get_system_time_msecs() - time)
```

elapsed times (tested 3 times):
22754 ms
23051 ms
20423 ms 

**C. insert_row() WITH transactions**

```gdscript
var time := OS.get_system_time_msecs()
db.query("BEGIN TRANSACTION;")
for i in range(0, NUMBER_OF_ROWS):
    db.insert_row(table_name, {"id": i, "number": randi()})
db.query("END TRANSACTION;")
print(OS.get_system_time_msecs() - time)
```

elapsed times (tested 3 times):
2893 ms
2751 ms
2838 ms 

**D. insert_rows() WITH transactions**

```gdscript
var time := OS.get_system_time_msecs()
var row_array := []
for i in range(0, NUMBER_OF_ROWS):
    row_array.append({"id": i, "number": randi()})
db.query("BEGIN TRANSACTION;")
db.insert_rows(table_name, row_array)
db.query("END TRANSACTION;")
print(OS.get_system_time_msecs() - time)
```

elapsed times (tested 3 times):
2718 ms
2695 ms
2958 ms 

So yeah, with transactions this is significantly faster!
However, there's no noticeable difference between `insert_row()` and `insert_rows()` which is unfortunate...
Guess it's only still nice to have for convenience 🤷‍♂️ 

Also when the user tries to start a transaction inside a transaction he/she gets this handy message:
![Capture](https://user-images.githubusercontent.com/42484461/104523201-dd6e9c80-55fe-11eb-8902-598c27321520.PNG)


I'll add this to the `insert_rows()`-function as you suggested 👌 (and to the docs as well ofc.)

**Comment by CitrusWire** [comment]
Cool stuff. Be sure to add it for `delete_rows` and `update_rows` as well.

**Comment by 2shady4u** [comment]
I've added transactions to the `feature-add-blobs` branch (which might need to be renamed to `develop`):

08390ee9d320497bf81dd54574b40e2bc64c097c

So next release will have transactions by default for the suggested functions.

**Comment by 2shady4u** [comment]
Implemented/resolved in version 2.1 as found in the [releases](https://github.com/2shady4u/godot-sqlite/releases/tag/v2.1) tab.
